atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ReadSentinelHandler.java

```diff
@@ -0,0 +1,136 @@
+    private final OrphanedSentinelDeleter orphanedSentinelDeleter;
+
+    public ReadSentinelHandler(
+            TransactionKeyValueService transactionKeyValueService,
+            TransactionService transactionService,
+            TransactionReadSentinelBehavior readSentinelBehavior,
+            OrphanedSentinelDeleter orphanedSentinelDeleter) {
+        this.transactionKeyValueService = transactionKeyValueService;
+        this.transactionService = transactionService;
+        this.readSentinelBehavior = readSentinelBehavior;
+        this.orphanedSentinelDeleter = orphanedSentinelDeleter;
+    }
+
+    /**
+     * A sentinel becomes orphaned if the table has been truncated while an old value was being swept; there is a
+     * chance that we end up with a sentinel with no valid AtlasDB cell covering it. In this case, we ignore it.
+     */
+    public Set<Cell> findAndMarkOrphanedSweepSentinelsForDeletion(TableReference table, Map<Cell, Value> rawResults) {
+        Set<Cell> sweepSentinels = Maps.filterValues(rawResults, ReadSentinelHandler::isSweepSentinel)
+                .keySet();
+        if (sweepSentinels.isEmpty()) {
+            return sweepSentinels;
+        }
+
+        // for each sentinel, start at long max. Then iterate down with each found uncommitted value.
+        // if committed value seen, stop: the sentinel is not orphaned
+        // if we get back -1, the sentinel is orphaned
+        Map<Cell, Long> timestampCandidates = new HashMap<>(
+                transactionKeyValueService.getLatestTimestamps(table, Maps.asMap(sweepSentinels, x -> Long.MAX_VALUE)));
+        Set<Cell> actualOrphanedSentinels = new HashSet<>();
+
+        while (!timestampCandidates.isEmpty()) {
+            Map<SentinelType, Map<Cell, Long>> sentinelTypeToTimestamps = timestampCandidates.entrySet().stream()
+                    .collect(Collectors.groupingBy(
+                            entry -> entry.getValue() == Value.INVALID_VALUE_TIMESTAMP
+                                    ? SentinelType.DEFINITE_ORPHANED
+                                    : SentinelType.INDETERMINATE,
+                            Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+
+            Map<Cell, Long> definiteOrphans = sentinelTypeToTimestamps.get(SentinelType.DEFINITE_ORPHANED);
+            if (definiteOrphans != null) {
+                actualOrphanedSentinels.addAll(definiteOrphans.keySet());
+            }
+
+            Map<Cell, Long> cellsToQuery = sentinelTypeToTimestamps.get(SentinelType.INDETERMINATE);
+            if (cellsToQuery == null) {
+                break;
+            }
+            Set<Long> committedStartTimestamps = KeyedStream.stream(transactionService.get(cellsToQuery.values()))
+                    .filter(commit -> commit != null && commit != TransactionConstants.FAILED_COMMIT_TS)
+                    .keys()
+                    .collect(Collectors.toSet());
```

Comment:
feedback for next time: would've been useful to leave a message here highlighting the different between the new and old version and where we're fixing the bug.

Reply:
yep! will do - yeah the difference is indeed that `commit != null` alone which was previously used, is insufficient